### PR TITLE
ARROW-3477: [C++] fixes for 32 bit architectures

### PR DESCRIPTION
**Pull request by veprbl**

### CODE REVIEW

**Review by wesm — APPROVED** [review]
Looks ok to me. We ought to set up a Dockerfile where we can periodically check the 32-bit build. Could someone else take a look at this also?

**Comment by kszucs** [comment]
We should create a ticket for that

File: cpp/src/arrow/buffer-test.cc [hunk 1]
```diff
@@ -272,7 +273,10 @@ TYPED_TEST(TypedTestBuffer, ResizeOOM) {
   TypeParam buf;
   ASSERT_OK(AllocateResizableBuffer(0, &buf));
   ASSERT_OK(buf->Resize(100));
-  int64_t to_alloc = std::numeric_limits<int64_t>::max();
+  int64_t to_alloc = std::min<uint64_t>(std::numeric_limits<int64_t>::max(),
```
Comment (fsaintjacques) [inline cpp/src/arrow/buffer-test.cc]: `min<int64_t>` and not uint

File: cpp/src/arrow/builder.cc [hunk 2]
```diff
@@ -736,7 +736,7 @@ Status BooleanBuilder::AppendValues(const std::vector<bool>& values,
 
   int64_t i = 0;
   internal::GenerateBitsUnrolled(raw_data_, length_, length,
-                                 [values, &i]() -> bool { return values[i++]; });
+                                 [&values, &i]() -> bool { return values[i++]; });
```
Comment (fsaintjacques) [inline cpp/src/arrow/builder.cc]: Why?

File: cpp/src/arrow/builder.cc [hunk 3]
```diff
@@ -749,7 +749,7 @@ Status BooleanBuilder::AppendValues(const std::vector<bool>& values) {
...
```
Comment (fsaintjacques) [inline cpp/src/arrow/builder.cc]: Same

File: cpp/src/arrow/util/bit-util.h [hunk 4]
```diff
@@ -170,7 +170,7 @@ static inline int CountLeadingZeros(uint32_t value) {
 static inline int CountLeadingZeros(uint64_t value) {
 #if defined(__clang__) || defined(__GNUC__)
   if (value == 0) return 64;
-  return static_cast<int>(__builtin_clzl(value));
+  return static_cast<int>(__builtin_clzll(value));
```
Comment (fsaintjacques) [inline cpp/src/arrow/util/bit-util.h]: good catch.

File: cpp/src/arrow/memory_pool.cc [hunk 5]
```diff
@@ -44,7 +45,13 @@ namespace {
 // Allocate memory according to the alignment requirements for Arrow
 // (as of May 2016 64 bytes)
 Status AllocateAligned(int64_t size, uint8_t** out) {
-// TODO(emkornfield) find something compatible with windows
+  // TODO(emkornfield) find something compatible with windows
```
Comment (fsaintjacques) [inline cpp/src/arrow/memory_pool.cc]: I'm sceptical of this new check for performance reason. Maybe hide all of this under asserts such that it is avoided in production?

File: cpp/src/arrow/memory_pool.cc [hunk 6]
```diff
@@ -104,7 +111,14 @@ class DefaultMemoryPool : public MemoryPool {
   Status Reallocate(int64_t old_size, int64_t new_size, uint8_t** ptr) override {
 #ifdef ARROW_JEMALLOC
     uint8_t* previous_ptr = *ptr;
-    *ptr = reinterpret_cast<uint8_t*>(rallocx(*ptr, new_size, MALLOCX_ALIGN(kAlignment)));
+    if (new_size < 0) {
```
Comment (fsaintjacques) [inline cpp/src/arrow/memory_pool.cc]: Ditto.

File: cpp/src/arrow/memory_pool-test.h [hunk 7]
```diff
@@ -53,7 +54,10 @@ class TestMemoryPoolBase : public ::testing::Test {
     auto pool = memory_pool();
 
     uint8_t* data;
-    int64_t to_alloc = std::numeric_limits<int64_t>::max();
+    int64_t to_alloc = std::min<uint64_t>(std::numeric_limits<int64_t>::max(),
```
Comment (fsaintjacques) [inline cpp/src/arrow/memory_pool-test.h]: Same as previous.

**Review by fsaintjacques — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/buffer-test.cc (hunk 1 above)
Comment (veprbl) [reply cpp/src/arrow/buffer-test.cc]: No. The value of `std::numeric_limits<size_t>::max()` can be as big as `std::numeric_limits<uint64_t>::max()` which overflows `int64_t`.

File: cpp/src/arrow/builder.cc (hunk 2 above)
Comment (veprbl) [reply cpp/src/arrow/builder.cc]: I don't know. This is crashing in the release version when it tries to copy the "values" vector into closure. It's also a good change, we don't want to make a copy in the first place.

File: cpp/src/arrow/memory_pool.cc (hunk 5 above)
Comment (veprbl) [reply cpp/src/arrow/memory_pool.cc]: Do you have benchmarks to demonstrate the performance loss here? If yes, how do I run them?

File: cpp/src/arrow/builder.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/builder.cc]: Ah, nice catch indeed.

File: cpp/src/arrow/memory_pool.cc (hunk 5 above)
Comment (pitrou) [reply cpp/src/arrow/memory_pool.cc]: I think those checks are reasonable. The Arrow allocator will usally be called with large-ish sizes (i.e. we're not using it to allocate tons of tiny objects), so the check shouldn't be problematic.

**Review by pitrou — APPROVED** [review]
+1, LGTM

**Pull request closed without merging**